Add rust runtime

nhynes opened the pull request:
This PR adds a Rust runtime which is useful when creating statically linked TVM modules--either native or Wasm. Indeed, Rust has the best Wasm support. It's also good for SGX.

The current code supports both TVM and NNVM modules. The API around `Tensor` conversions to `DLTensor` is a bit awkward since `DLTensor` isn't owned, but it shouldn't be too hard to iron out.

There's one nit, though: the `dshape` field in the `Tensor` struct exists solely because TVM uses `u64` as the `shape_t` even when on 32-bit platforms. I'll make a PR for this eventually...

nhynes commented on src/runtime/rust/src/runtime/threading.rs:
```diff
@@ -0,0 +1,299 @@
+  if let Some(q) = SGX_QUEUES.lock().unwrap().pop_front() {
+    ThreadPool::run_worker(q);
+  }
+}
```
this function interacts with the C++ runtime dylib

nhynes commented on src/runtime/rust/src/runtime/array.rs:
```diff
@@ -0,0 +1,393 @@
+  pub(super) strides: Option<Vec<usize>>,
+  pub(super) byte_offset: isize,
+  pub(super) numel: usize,
+  pub(super) dshape: Vec<i64>,
```
this is the `dshape` mentioned in the PR description. When the target arch is 32-bit, `usize` is 32 bits. SInce the TVM module wants a 64-bit shape array, there needs to be an owned copy stored somewhere (i.e. in `dshape`).

nhynes commented on src/runtime/rust/.rustfmt.toml:
```diff
@@ -0,0 +1,57 @@
+report_todo = "Never"
+report_fixme = "Never"
+ignore = []
+verbose_diff = false
```
this is the default `rustfmt.toml` produced by `rustfmt --dump-default-config`

tqchen commented:
@jroesch @ehsanmok can you help review this?

tqchen commented:
Let us directly put it under tvm/rust

tqchen commented:
+1 for docs and comments, I would recommend having one example for users to get started, let us make sure all the public API are documented and commented, that is what we need reviews for :)

also created an issue following up the discussion https://github.com/dmlc/tvm/issues/1601

ehsanmok commented:
@tqchen yes, it was confusing to me that this PR adds more and is different from what already exists for "runtime frontend" support that I had in mind which I think is the intention for v0.5 roadmap.

nhynes commented:
@ehsanmok 

> What's the clear and correct way of compiling tvm-rs?

`cargo build`. The error you were getting before was due to the `build.rs` being ignored by git.

> What's wrong with having the exact runtime API support as Java

Nothing at all. That's what we're working on: a unified Rust crate-of-crates which supports both frontend and "backend" Rust. Currently, having a Rust backend is useful for the emerging technologies of Wasm and SGX. These are effectively CPU targets, so that's where the effort has gone.

> documentation and example repo

The tests could easily be made into docs. I'll wait for @jroesch's reviews before diving into docs.

jroesch commented on rust/.travis.yml:
```diff
@@ -0,0 +1,5 @@
+language: rust
+rust:
+  - nightly
```
Do we need to be on Rust nightly? may be wishful thinking but I was hoping after almost 30 releases people would be using stable ;)

nhynes replied:
> people would be using stable

Ha, good one! :)

Unfortunately, this crate requires `#![feature(allocator_api, box_syntax, fn_traits, try_from, unboxed_closures)]` (`box_syntax` is _totally_ necessary).

jroesch commented on rust/src/runtime/array.rs:
```diff
@@ -0,0 +1,393 @@
+  }
+}
+
+fn tensor_from_array_storage<'a, 's, T, D: ndarray::Dimension>(
```
is it possible to make these static methods?

jroesch commented on rust/src/runtime/array.rs:
```diff
@@ -0,0 +1,393 @@
+  }
+}
+
+macro_rules! impl_tensor_from_ndarray {
```
These could probably use some explanation.

jroesch commented on rust/src/runtime/graph.rs:
```diff
@@ -0,0 +1,408 @@
+  DLDataTypeCode_kDLFloat, DLDataTypeCode_kDLInt, DLDataTypeCode_kDLUInt, DLTensor,
+};
+
+const NDARRAY_MAGIC: u64 = 0xDD5E40F096B4A13F; // Magic number for NDArray file
```
These need a deeper explanation.

jroesch commented:
Just did a quick pass before bed, code looks good, but could use lots of docs explaining it. I think it would be very hard for someone who didn't know the existing runtime to understand. I'll take another stab tomorrow or after you do docs? just let me know what would be more useful.

nhynes commented:
thanks for your review @jroesch! In the interest of making efficient use of time, I'll make a docs pass before you do further review.

ehsanmok commented on src/runtime/rust/src/runtime/array.rs:
```diff
@@ -0,0 +1,393 @@
+}
+
+impl<'a> Tensor<'a> {
+  pub fn shape(&self) -> Vec<usize> {
```
Inlinable with `#[inline]` and the like methods.

ehsanmok replied:
Partially agree! while it's debatable, my reference is bluss's ndarray and the like libs he's wrote with careful inlining. I'm not familiar with wasm though.

nhynes replied:
I'm not sure that it's a debate: LTO is unconditionally better. The ndarray crate is relatively old--perhaps bluss was using best bractices from a time when the Rust linker was less sophisticated.

References:

* https://stackoverflow.com/questions/37639276/when-should-inline-be-used-in-rust
* https://stackoverflow.com/questions/1759300/when-should-i-write-the-keyword-inline-for-a-function-method
* https://stackoverflow.com/questions/7046547/link-time-optimization-and-inline
* https://stackoverflow.com/questions/5948703/difference-in-inlining-functions-by-compiler-or-linker

ehsanmok replied:
Thanks for the links! I'll look into them. The major source of (anything in fact and) inlining is in std, [for example](https://doc.rust-lang.org/src/core/option.rs.html#191). I think, if in doubt in general, then skip it for now.

ehsanmok commented on src/runtime/rust/src/runtime/array.rs:
```diff
@@ -0,0 +1,393 @@
+}
+
+impl DataType {
+  fn itemsize(&self) -> usize {
```
The same `#[inline]`. (Not repeating for the like)

nhynes commented on src/runtime/rust/src/runtime/allocator.rs:
```diff
@@ -0,0 +1,46 @@
+const DEFAULT_ALIGN_BYTES: usize = 4;
+
+#[derive(PartialEq, Eq)]
+pub struct Allocation {
```
This isn't actually an allocator--it's just a way to get bytes out of the global allocator in a less-unstable way.

nhynes commented:
@jroesch I added some docs. Hopefully they're sufficient for your critical review.

@ehsanmok Thanks for the feedback. I addressed most of it. I'll attend to the other comments when we're closer to merging.

nhynes commented:
[Rendered docs](http://nhynes.github.io/tvm_rs/tvm/index.html)

ehsanmok commented:
@nhynes thanks! a couple of points/questions:

1) Besides adding a separate _example repo_ with a couple of end-to-end use cases (cleaner version of integrations tests maybe?), I suggest adding more docs specially in `src/lib.rs` describing what tvm is, links to tvm.ai, more tutorials etc. and what this runtime lib provides. 

2) For cleaner rendering of docs, I suggest to not include the dependencies and/or use `#[doc(hidden)]` to control the visibilities of some parts of this lib. For example, consider hiding the constants in `tvm::ffi::runtime`.

3) Question: (to @tqchen as well) are we considering to support rust as a backend replacement of c++ ever? if so, I hardly think it'd be a good plausible idea and if not, then shall we get rid of the separate runtime module namespace in this PR and simply put everything there in the root lib?

nhynes commented:
> place everything in `tvm::runtime` as simply in root `tvm`

What you're working on isn't a runtime, though. It'd still go in the `tvm/rust` directory but have a different crate. We could have `rust/tvm/runtime` be `tvm-runtime` and your frontend, simply `tvm`.

> rust compiler backend

Not until rust has good python FFI support. That's the main benefit of C[++]

tqchen commented:
@ehsanmok @jroesch @nhynes it is a good time to revive this thread, let us add follow up comments and work to merge this in

ehsanmok commented:
@tqchen the last number of changes addressed some of my concerns and comments above. Besides improving docs and some other smaller things (like `#[inline]` attributes) that can be done later, my main concern is related to [#1226](https://github.com/dmlc/tvm/issues/1226) and seems allowing it in the [build](https://github.com/dmlc/tvm/pull/1597/files#diff-6228dfbca0bfcb4a4662ce8cc1ccf963R8). If @nhynes thinks it was already resolved, then I think there shouldn't be a major blocking issue to merge. If not, then at least we can remove the feature and fix it later!

nhynes commented:
> my main concern is related to #1226 and seems allowing it in the build

I haven't looped around to fixing that, but since the rust runtime is mostly useful for Wasm (single threaded) and SGX (which doesn't manage its own threads), it's not a blocker.

ehsanmok commented:
@nhynes Ok! So we shouldn't allow the related `par-launch-alloc` feature in master until it is resolved.

nhynes commented:
Actually, I just fixed the bug. Turns out the issue was that the temporal workspaces only work correctly if they have the same alignment as the dtype they're supposed to hold. Go figure.

ehsanmok commented:
Great! noticed the added `Send+Sync`. This should be good to go then.

tqchen commented:
Let us push to get this in, @nhynes please send a separate PR to upstream to update the Dockerfile.demo_cpu with necessary rust env.

@nhynes I take a quick look, although I am not rust expert, it seems to me that the documentation are still sparse. Please at least document
- The tvm runtime module (e.g. This is a rust implementation of tvm runtime, can be used for SGX, WASM etc...)
- The user-facing API functions (borrow docs from TVM API)
- Some of the subtle implementation points that @jroesch mentioned.

After this is done and test get passed we can merge it in and followup with @ehsanmok on rust frontend

nhynes commented:
> documentation are still sparse

The convention is to use the auto-generated docs with manual docs only when necessary. Here's a [link to the current docs](http://nhynes.github.io/tvm-rs/tvm/runtime/index.html). It's also populated with [examples](http://nhynes.github.io/tvm-rs/tvm/runtime/struct.GraphExecutor.html).

I'll add a top-level module doc, add a dockerfile config, and address those code review comments.

nhynes commented:
> `Dockerfile.demo_cpu`

should I also update CI? Rust builds might be a bit flaky since this crate needs to use nightly, though

tqchen commented:
Oh, yes, I meant to say ci_cpu. Using nightly is fine. However, note that we build the docker image infrequently, which means we will be stuck to a certain version of nightly. Using a stable version  or hastag is preferred

tqchen commented:
Please add rust testcases to Jenkinsfile to here https://github.com/dmlc/tvm/blob/master/Jenkinsfile#L141

nhynes commented:
the testcases are added here https://github.com/nhynes/tvm/pull/1

I'll make a PR for that here when this branch and #1825 are merged

tqchen commented:
Thanks @jroesch @nhynes @ehsanmok , this is now merged
